File: day2.js
var fs = require('fs');

function surface(l, h, w) {
  return 2*l*w + 2*w*h + 2*h*l;
}

function slack(l, h, w) {
  return Math.min(l*w, l*h, w*h);
}

function bow(l, h, w) {
  return l*w*h;
}

function ribbonForBox(l, h, w) {
  return 2 * (l + h + w - Math.max(l, h, w));
}

function sqFtForPackage(arr) {
  var l = parseInt(arr[0]);
  var h = parseInt(arr[1]);
  var w = parseInt(arr[2]);
  return surface(l, h, w) + slack(l, h, w);
}

function totalRibbon(arr) {
  var l = parseInt(arr[0]);
  var h = parseInt(arr[1]);
  var w = parseInt(arr[2]);
  return bow(l, h, w) + ribbonForBox(l, h, w);
}

function calculateSqFtForAllPackages(file) {
  var packages;
  var wrappingPaper = 0;
  var ribbon = 0;

  fs.readFile(file, 'utf8', function(err, data) {
    if (err) {
      return console.log(err);
    }

    var rawDataArray = data.split("\n");

    packages = rawDataArray.map(function(package) {
      return package.split("x");
    });

    packages.pop();

    packages.forEach(function(package) {
      wrappingPaper += sqFtForPackage(package);
      ribbon += totalRibbon(package);
    });

    console.log("Square Feet of Wrapping Paper Required: " + wrappingPaper);
    console.log("Length of Ribbon Required: " + ribbon + " feet");
  });
}

calculateSqFtForAllPackages('packages.txt');

File: day2/day2.rb
paper_needed = 0
ribbon_needed = 0

File.open('packages.txt', 'r') do |f|
  f.each_line do |line|
    arr = line.split("x").map(&:to_i)
    side_a = 2 * arr[0] * arr[1]
    side_b = 2 * arr[1] * arr[2]
    side_c = 2 * arr[2] * arr[0]
    area = [side_a, side_b, side_c]
    # Wrapping Paper
    paper_needed += area.push(area.min / 2).inject(:+)

    # Ribbon to Wrap around package
    ribbon_needed += (arr.inject(:+) - arr.max) * 2

    # Add Bow
    ribbon_needed += (arr.inject(:*))
  end
end

puts "Part 1: #{paper_needed}"
puts "Part 2: #{ribbon_needed}"
